File: modules/expo-auto-attendance/android/src/main/java/expo/modules/autoattendance/GeofenceEventBus.kt
package expo.modules.autoattendance

/**
 * In-process bridge between [GeofenceBroadcastReceiver] (which runs outside the
 * React Native context, possibly with no JS loaded at all) and
 * [ExpoAutoAttendanceModule] (which forwards events to JS listeners).
 *
 * The receiver persists every event via [GeofenceStore] before emitting, so nothing
 * is lost when the app process was woken up just for the broadcast and no listener
 * is attached.
 */
object GeofenceEventBus {
  fun interface Listener {
    fun onEvent(eventName: String, payload: Map<String, Any?>)
  }

  @Volatile
  private var listener: Listener? = null

  fun attach(listener: Listener) {
    this.listener = listener
  }

  fun detach() {
    listener = null
  }

  fun emit(eventName: String, payload: Map<String, Any?>) {
    listener?.onEvent(eventName, payload)
  }
}

File: modules/expo-auto-attendance/ios/GeofenceEventBus.swift
import Foundation

/// Event names shared between the manager and the Expo module definition.
enum GeofenceEvents {
  static let enter = "onGeofenceEnter"
  static let exit = "onGeofenceExit"
  static let error = "onError"
}

/**
 In-process bridge between `GeofenceManager` (whose Core Location delegate can
 fire while no JS is loaded, e.g. after a background relaunch) and
 `ExpoAutoAttendanceModule`, which forwards events to JS listeners.

 Events are persisted via `GeofenceStore` before being emitted, so nothing is
 lost when no listener is attached — mirroring the Android `GeofenceEventBus`.
 */
final class GeofenceEventBus {
  static let shared = GeofenceEventBus()

  var listener: ((_ eventName: String, _ payload: [String: Any?]) -> Void)?

  private init() {}

  func emit(_ eventName: String, _ payload: [String: Any?]) {
    listener?(eventName, payload)
  }
}
